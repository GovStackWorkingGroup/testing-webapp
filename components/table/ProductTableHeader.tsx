import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import {
  RiArrowDownSFill,
  RiArrowUpSFill,
} from 'react-icons/ri';
import classNames from 'classnames';
import { ColumnSortType, SortFieldType, TableSortByType } from './types';

type Props = {
  handleSorting: (tableSortProperties: TableSortByType) => void;
};

const ProductTableHeader = ({ handleSorting }: Props) => {
  const [softwareSort, setSoftwareSort] = useState<ColumnSortType>({ field: 'software',order: null });
  const [dateSort, setDateSort] = useState<ColumnSortType>({ field: 'date', order: null });
  const [compatibilitySort, setCompatibilitySort] = useState<ColumnSortType>({ field: 'compatibility', order: null });

  const { formatMessage } = useIntl();
  const format = useCallback(
    (id: string) => formatMessage({ id }),
    [formatMessage]
  );

  useEffect(() => {
    handleSorting({ software: softwareSort, date: dateSort, compatibility: compatibilitySort });
  }, [softwareSort, dateSort, compatibilitySort]);

  const handleSortingChange = (field: SortFieldType) => {
    if (field === 'software') {
      setSoftwareSort({
        field: 'software',
        order: softwareSort.order === 'asc' ? 'desc' : (softwareSort.order === 'desc' ? null : 'asc'),
      });
    } else if (field === 'date') {
      setDateSort({
        field: 'date',
        order: dateSort.order === 'asc' ? 'desc' : (dateSort.order === 'desc' ? null : 'asc'),
      });
    } else if (field === 'compatibility') {
      setCompatibilitySort({
        field: 'compatibility',
        order: compatibilitySort.order === 'asc' ? 'desc' : (compatibilitySort.order === 'desc' ? null : 'asc'),
      });
    }
  };

  return (
    <div className='product-table-header'>
      <div></div>
      <div
        onClick={() => handleSortingChange('software')}
        data-testid='header-sort-software'
        className='cursor-pointer'
      >
        <p>{format('software_name.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', { 'sort-icon-active': softwareSort.order === 'asc' })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', { 'sort-icon-active': softwareSort.order === 'desc' })}
          />
        </div>
      </div>
      <div>
        <p>{format('building_block.plural.label')}</p>
      </div>
      <div
        onClick={() => handleSortingChange('date')}
        data-testid='header-sort-update'
        className='cursor-pointer'
      >
        <p>{format('table.last_update.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', { 'sort-icon-active': dateSort.order === 'asc' })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', { 'sort-icon-active': dateSort.order === 'desc' })}
          />
        </div>
      </div>
      <div
        onClick={() => handleSortingChange('compatibility')}
        data-testid='header-sort-compatibility'
        className='cursor-pointer'
      >
        <p>{format('table.overall_compatibility.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', { 'sort-icon-active': compatibilitySort.order === 'asc' })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', { 'sort-icon-active': compatibilitySort.order === 'desc' })}
          />
        </div>
      </div>
      <div></div>
    </div>
  );
};

export default ProductTableHeader;
