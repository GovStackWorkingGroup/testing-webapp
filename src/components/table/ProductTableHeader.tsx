import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import {
  RiArrowDownSFill,
  RiArrowUpSFill,
} from 'react-icons/ri';
import classNames from 'classnames';
import { SoftwaresTablColumnSortType, SortSoftwareTableFieldType, SoftwaresTableSortByType } from './types';

type Props = {
  handleSorting: (tableSortProperties: SoftwaresTableSortByType) => void;
};

const ProductTableHeader = ({ handleSorting }: Props) => {
  const [softwareSort, setSoftwareSort] = useState<SoftwaresTablColumnSortType>({ field: 'testApp',order: null });
  const [dateSort, setDateSort] = useState<SoftwaresTablColumnSortType>({ field: 'lastUpdate', order: null });
  const [compatibilitySort, setCompatibilitySort] = useState<SoftwaresTablColumnSortType>(
    { field: 'overallCompatibility', order: null }
  );

  const { formatMessage } = useIntl();
  const format = useCallback(
    (id: string) => formatMessage({ id }),
    [formatMessage]
  );

  useEffect(() => {
    handleSorting({ testApp: softwareSort, lastUpdate: dateSort, overallCompatibility: compatibilitySort });
  }, [softwareSort, dateSort, compatibilitySort]);

  const handleSortingChange = (field: SortSoftwareTableFieldType) => {
    if (field === 'testApp') {
      setSoftwareSort({
        field: 'testApp',
        order: softwareSort.order === 'asc' ? 'desc' : (softwareSort.order === 'desc' ? null : 'asc'),
      });
    } else if (field === 'lastUpdate') {
      setDateSort({
        field: 'lastUpdate',
        order: dateSort.order === 'asc' ? 'desc' : (dateSort.order === 'desc' ? null : 'asc'),
      });
    } else if (field === 'overallCompatibility') {
      setCompatibilitySort({
        field: 'overallCompatibility',
        order: compatibilitySort.order === 'asc' ? 'desc' : (compatibilitySort.order === 'desc' ? null : 'asc'),
      });
    }
  };

  return (
    <div className='product-table-header'>
      <div></div>
      <div
        onClick={() => handleSortingChange('testApp')}
        data-testid='header-sort-software'
        className='cursor-pointer'
      >
        <p>{format('software_name.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', { 'sort-icon-active': softwareSort.order === 'asc' })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', { 'sort-icon-active': softwareSort.order === 'desc' })}
          />
        </div>
      </div>
      <div>
        <p>{format('building_block.plural.label')}</p>
      </div>
      <div
        onClick={() => handleSortingChange('lastUpdate')}
        data-testid='header-sort-update'
        className='cursor-pointer'
      >
        <p>{format('table.last_update.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', { 'sort-icon-active': dateSort.order === 'asc' })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', { 'sort-icon-active': dateSort.order === 'desc' })}
          />
        </div>
      </div>
      <div
        onClick={() => handleSortingChange('overallCompatibility')}
        data-testid='header-sort-compatibility'
        className='cursor-pointer'
      >
        <p>{format('table.overall_compatibility.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', { 'sort-icon-active': compatibilitySort.order === 'asc' })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', { 'sort-icon-active': compatibilitySort.order === 'desc' })}
          />
        </div>
      </div>
      <div></div>
    </div>
  );
};

export default ProductTableHeader;
