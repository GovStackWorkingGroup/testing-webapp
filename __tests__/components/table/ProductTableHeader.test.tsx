import { fireEvent } from '@testing-library/react';
import ProductTableHeader from '../../../components/table/ProductTableHeader';
import { render } from '../../test-utils/test-utils';

describe('Unit tests for ProductTableHeader component:', () => {
  const SORT_SOFTWARE_TEST_ID = 'header-sort-software';
  const SORT_LAST_UPDATE_TEST_ID = 'header-sort-update';
  const SORT_COMPATIBILITY_TEST_ID = 'header-sort-compatibility';

  it('should render the ProductTableHeader component and match snapshot', () => {
    const { container } = render(<ProductTableHeader handleSorting={jest.fn()} />);

    expect(container).toMatchSnapshot();
  });

  it('should call the sorting function after clicks on the Software header or sort icon', () => {
    const mockedHandleSorting = jest.fn();
    const { getByTestId } = render(<ProductTableHeader handleSorting={mockedHandleSorting} />);

    expect(mockedHandleSorting).not.toHaveBeenCalled();

    fireEvent.click(getByTestId(SORT_SOFTWARE_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(1);
    expect(mockedHandleSorting).toHaveBeenCalledWith('software', 'asc');

    fireEvent.click(getByTestId(SORT_SOFTWARE_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(2);
    expect(mockedHandleSorting).toHaveBeenCalledWith('software', 'desc');
  });

  it('should call the sorting function after clicks on the Last update header or sort icon', () => {
    const mockedHandleSorting = jest.fn();
    const { getByTestId } = render(<ProductTableHeader handleSorting={mockedHandleSorting} />);

    expect(mockedHandleSorting).not.toHaveBeenCalled();

    fireEvent.click(getByTestId(SORT_LAST_UPDATE_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(1);
    expect(mockedHandleSorting).toHaveBeenCalledWith('date', 'asc');

    fireEvent.click(getByTestId(SORT_LAST_UPDATE_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(2);
    expect(mockedHandleSorting).toHaveBeenCalledWith('date', 'desc');
  });

  it('should call the sorting function after clicks on the Compatibility header or sort icon', () => {
    const mockedHandleSorting = jest.fn();
    const { getByTestId } = render(<ProductTableHeader handleSorting={mockedHandleSorting} />);

    expect(mockedHandleSorting).not.toHaveBeenCalled();

    fireEvent.click(getByTestId(SORT_COMPATIBILITY_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(1);
    expect(mockedHandleSorting).toHaveBeenCalledWith('compatibility', 'asc');

    fireEvent.click(getByTestId(SORT_COMPATIBILITY_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(2);
    expect(mockedHandleSorting).toHaveBeenCalledWith('compatibility', 'desc');
  });
});
