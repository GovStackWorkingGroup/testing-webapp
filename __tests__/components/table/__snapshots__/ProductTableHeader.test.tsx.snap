// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for ProductTableHeader component: should render the ProductTableHeader component and match snapshot 1`] = `
<div>
  <div
    class="product-table-header"
  >
    <div />
    <div>
      <p>
        Product Name
      </p>
    </div>
    <div>
      <p>
        Building Blocks
      </p>
    </div>
    <div>
      <p>
        Last Update
      </p>
    </div>
    <div>
      <p>
        Overall Compatibility
      </p>
    </div>
    <div />
  </div>
</div>
`;
